File: leetcode/rb/19.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @param {Integer} n
# @return {ListNode}
def remove_nth_from_end(head, n)
  dummy = ListNode.new(nil)
  dummy.next = head

  fast, slow = dummy, dummy

  n.times { fast = fast.next }

  while fast.next
    fast = fast.next
    slow = slow.next
  end

  slow.next = slow.next.next

  return dummy.next
end

File: leetcode/py/19.py
class Solution:
    def removeNthFromEnd(self, head, n):
        if not head:
            return None
        dummy = ListNode(None)
        dummy.next = head

        fast, slow = dummy, dummy

        for _ in range(n):
            fast = fast.next

        while fast.next:
            fast = fast.next
            slow = slow.next

        slow.next = slow.next.next

        return dummy.next
